src/java/org/apache/cassandra/net/BufferPoolAllocator.java

```diff
@@ -92,17 +98,37 @@ void release()
     public static class Wrapped extends UnpooledUnsafeDirectByteBuf
     {
         private ByteBuffer wrapped;
+        private boolean returnToPool = true;
 
-        Wrapped(BufferPoolAllocator allocator, ByteBuffer wrap)
+        Wrapped(BufferPoolAllocator allocator, ByteBuffer wrap, int maxCapacity)
         {
-            super(allocator, wrap, wrap.capacity());
+            super(allocator, wrap, max(wrap.capacity(), maxCapacity));
             wrapped = wrap;
         }
 
+        @Override
+        public ByteBuf capacity(int newCapacity)
+        {
+            if (newCapacity == capacity())
+                return this;
+
+            // resizing doesn't use the pool
+            ByteBuf newBuffer = super.capacity(newCapacity);
```

Comment:
not sure why we need to use `super.capacity(newCapacity);` which doesn't use our buffer pool. 

is it possible to directly allocate new capacity from buffer pool? and then copy old buffer to new buffer. at the end, release the old buffer. So we don't have to keep track of `returnToPool` which defeats the purpose of reducing fragmented buffer allocation.

wdyt? @grighetto @bereng

Reply:
@jsonstack I went that route initially, but there's a problem with that approach, we pass the initial buffer to the constructor of `UnpooledUnsafeDirectByteBuf` and it keeps a reference to that internally:

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L42

We would have to call `setByteBuffer` to replace that object, but it is not public:

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L114

If we don't replace that internal buffer, there will be mismatches because some internal functions use it.
`super.capacity(newCapacity)` already does the right things, creates a new buffer, copies over the bytes, releases the old and replaces that reference.

> So we don't have to keep track of returnToPool which defeats the purpose of reducing fragmented buffer allocation

Right, I thought about that too and I agree, but for most cases it will behave as before (it has been working with fixed-length buffers up until now). Increasing the buffer capacity should be an edge case as the one reported in the ticket.

Reply:
That is the worry I had and I raised in my previous comment. You have to check Netty code to see if everything is consistent with Netty.

Reply:
> Increasing the buffer capacity should be an edge case as the one reported in the ticket

I think that's the key here. Making pooled, unpooled and netty work together nicely seems to be the problem.

BTW @grighetto did you check if upon resizing we should be checking any pool/general config memory size limit?

Reply:
one more issues with current implementation: when `capacity` is called, the `newBuffer` is never deallocated on buf `release` unless `capacity` is called again. But again, the latest `newBuffer` with the latest capacity is not deallocated.

Reply:
the `newBuffer` created by `super.capacity()` is not freed by `freeDirect()`, would be nice to handle it in `deallocate()`

Reply:
I have the same questions as @bereng here. I want to do this to match the 3.11 implementation if possible with minimum hurdle.
Not saying @grighetto didn't do it, he did a lot and great job, I am just confirming those details.

Reply:
If in 3.11 it is always used by Netty Integer.MAX_VALUE I would go for that path probably. Confirming with @grighetto the details in Slack